internal/security/bootstrapper/command/install/command.go

```diff
@@ -0,0 +1,144 @@
+	}
+	var dummy string
+
+	flagSet := flag.NewFlagSet(CommandName, flag.ContinueOnError)
+	flagSet.StringVar(&dummy, "confdir", "", "") // handled by bootstrap; duplicated here to prevent arg parsing errors
+
+	err := flagSet.Parse(args)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse command: %s: %w", strings.Join(args, " "), err)
+	}
+
+	return &cmd, err
+}
+
+func (c *cmd) Execute() (statusCode int, err error) {
+	c.loggingClient.Info(fmt.Sprintf("Security bootstrapper running %s", CommandName))
+
+	bootstrapServer := tcp.NewTcpServer()
+	c.loggingClient.Debug(fmt.Sprintf("init phase: attempts to start up the listener on bootstrap port: %d",
+		c.config.StageGate.BootstrapPort))
+
+	gatingSemaphorePort(bootstrapServer, c.config.StageGate.BootstrapPort, c.loggingClient,
+		"Raised bootstrap semaphore for secure bootstrapping")
+
+	// wait on for others to be done: each of tcp dialers is a blocking call
+	c.loggingClient.Debug("Waiting on dependent semaphores required to raise the ready-to-run semaphore ...")
+	if err := tcp.DialTcp(c.config.StageGate.ConsulHost, c.config.StageGate.ConsulReadyPort, c.loggingClient); err != nil {
+		c.loggingClient.Error(err.Error())
+	}
+	c.loggingClient.Info("Consul is ready")
+
+	if err := tcp.DialTcp(c.config.StageGate.PostgresHost, c.config.StageGate.PostgresReadyPort, c.loggingClient); err != nil {
+		c.loggingClient.Error(err.Error())
+	}
+	c.loggingClient.Info("Postgres is ready")
+
+	if err := tcp.DialTcp(c.config.StageGate.RedisHost, c.config.StageGate.RedisReadyPort, c.loggingClient); err != nil {
+		c.loggingClient.Error(err.Error())
+	}
+	c.loggingClient.Info("Redis is ready")
+
+	// Reached ready-to-run phase
+	c.loggingClient.Debug(fmt.Sprintf("ready-to-run phase: attempts to start up the listener on ready-to-run port: %d",
+		c.config.StageGate.ReadyToRunPort))
+
+	readToRunServer := tcp.NewTcpServer()
+
+	gatingSemaphorePort(readToRunServer, c.config.StageGate.ReadyToRunPort, c.loggingClient,
+		"Raised ready-to-run semaphore for secure bootstrapping")
+
+	// keep running until ctx done
+	c.waitGroup.Add(1)
+	go func() {
+		defer c.waitGroup.Done()
+
+		<-c.cntx.Done()
+		c.loggingClient.Info("security bootstrapper install finished")
+	}()
+
+	return
+}
+
+func (c *cmd) GetCommandName() string {
+	return CommandName
+}
+
+func gatingSemaphorePort(tcpServer *tcp.TcpServer, portNum int, lc logger.LoggingClient, raisedMsg string) {
+	// in a separated goroutine so that it won't block the main thread
+	go func() {
+		if err := tcpServer.StartListener(portNum, lc, ""); err != nil {
+			// listener is blocking forever until some internal critical error happens
+			lc.Error(err.Error())
+			os.Exit(1)
+		}
+	}()
+
+	for {
+		// looping until listener accepts, in which case the port semaphore is raised
+		accepted := tcpServer.GetAcceptStatus()
+		if accepted {
+			lc.Info(raisedMsg)
+			break
+		}
+
+		// give a little breathing room
+		time.Sleep(100 * time.Millisecond)
+	}
+}
```

Comment:
Why is this `accepted` needed? The listener will keep going for other service to connect.

Reply:
because want to have a log message to indicate it's opening the gate and i think that info message is important

Reply:
Still doesn't make since to me . We have opened the gate by simply listening to the port. The first service to connect isn't opening the gate.

Reply:
@lenny-intel  why not? it is confusing you said so-  please see the ADR for definition of that: https://github.com/edgexfoundry/edgex-go/blob/master/internal/security/config/command/help/command.go#L36-L46   the gate is raised as soon as all secret-realted stateful services are ready; in this case, we already waited for `RedisReady`, `ConsulReady`, and `PostgresReady` before connecting to this port...

Reply:
as discussed yesterday, this whole for-loop block is gone.